### PR TITLE
Use injector configmap for sidecar version mismatch analysis

### PR DESCRIPTION
**Pull request by jasonwzm**
fix https://github.com/istio/istio/issues/20858

### CODE REVIEW

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go [hunk 1]
```diff
@@ -96,43 +109,24 @@ func (a *VersionAnalyzer) Analyze(c analysis.Context) {
...
 }
 
-// tryReturnSidecarInjectorVersion returns an empty string if the pod is not
-// the sidecar injector; otherwise the version of the injector image is
-// returned.
-func tryReturnSidecarInjectorVersion(p *v1.Pod) string {
-	if p.Labels["app"] != sidecarInjectorName {
+// getSidecarinjectionVersion retrieves the injection version defined in the
+// sidecar injector configuration.
+func getSidecarinjectionVersion(cm *v1.ConfigMap) string {
```
Comment (howardjohn) [inline galley/pkg/config/analysis/analyzers/injection/injection-version.go]: nit: tag is not the same as version

Comment (jasonwzm) [reply galley/pkg/config/analysis/analyzers/injection/injection-version.go]: done

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go [hunk 2]
```diff
@@ -49,12 +51,31 @@ func (a *VersionAnalyzer) Metadata() analysis.Metadata {
...
 			collections.K8SCoreV1Pods.Name(),
+			collections.K8SCoreV1Configmaps.Name(),
 		},
 	}
 }
 
 // Analyze implements Analyzer.
 func (a *VersionAnalyzer) Analyze(c analysis.Context) {
+	var injectionVersion string
+	c.ForEach(collections.K8SCoreV1Configmaps.Name(), func(r *resource.Instance) bool {
```
Comment (selmanj) [inline galley/pkg/config/analysis/analyzers/injection/injection-version.go]: You can use `c.Find(c collection.Name, name resource.FullName)` to look up the resource by name, which makes this more readable

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go [hunk 3]
```diff
@@ -49,12 +51,31 @@ func (a *VersionAnalyzer) Metadata() analysis.Metadata {
...
+		if r.Metadata.FullName.Namespace.String() == "istio-system" &&
```
Comment (selmanj) [inline galley/pkg/config/analysis/analyzers/injection/injection-version.go]: We should add a TODO here that we need some way to specify the namespace the injector is running in.

In theory you could look for it across all namespaces, but I think it's fine to rely on istio-system right now.

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go [hunk 4]
```diff
@@ -96,43 +110,24 @@ func (a *VersionAnalyzer) Analyze(c analysis.Context) {
 			if v == "" {
 				continue
 			}
-			// Note the pod/version to check later after we've collected all injector versions.
-			podVersions = append(podVersions, podVersion{
-				Resource:     r,
-				ProxyVersion: v})
 
+			if v != injectionVersion {
```
Comment (selmanj) [inline galley/pkg/config/analysis/analyzers/injection/injection-version.go]: I was going to add a comment saying "Don't forget to handle the case where the "sidecar.istio.io/proxyImage" doesn't match the running pod, but now that I think about it, that shouldn't happen - if the annotation is applied then the injector runs immediately (as it's on the resource that's being mutated by the webhook).

Does that sound right to you?

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go (hunk 2 above)
Comment (jasonwzm) [reply galley/pkg/config/analysis/analyzers/injection/injection-version.go]: Done. thanks!

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go (hunk 3 above)
Comment (jasonwzm) [reply galley/pkg/config/analysis/analyzers/injection/injection-version.go]: Added a comment there. The current implementation is using istio-system.

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go (hunk 4 above)
Comment (jasonwzm) [reply galley/pkg/config/analysis/analyzers/injection/injection-version.go]: Yes, the image is updated to the image specified in the annotation. Plus we are skipping pods with that annotation in this analyzer.

File: galley/pkg/config/analysis/analyzers/injection/injection-version.go [hunk 5]
```diff
@@ -59,18 +59,14 @@ func (a *VersionAnalyzer) Metadata() analysis.Metadata {
...
 func (a *VersionAnalyzer) Analyze(c analysis.Context) {
 	var injectionVersion string
-	c.ForEach(collections.K8SCoreV1Configmaps.Name(), func(r *resource.Instance) bool {
-		if r.Metadata.FullName.Namespace.String() == "istio-system" &&
-			r.Metadata.FullName.Name.String() == sidecarInjectorConfigName {
-			cm := r.Message.(*v1.ConfigMap)
 
-			// The image tag name currently serves as the version for the injector.
-			injectionVersion = getIstioImageTag(cm)
+	// relying on just istio-system namespace for now as it align with the implementation.
```
Comment (selmanj) [inline galley/pkg/config/analysis/analyzers/injection/injection-version.go]: nit: this comment could be approved. which implementation? what do you mean by 'align'?

Comment (jasonwzm) [reply galley/pkg/config/analysis/analyzers/injection/injection-version.go]: done. Made a more clear.

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 6]
```diff
@@ -59,15 +59,15 @@ messages:
...
-    description: "The version of the Istio proxy running on the pod does not match the version used by the istio injector."
-    template: "The version of the Istio proxy running on the pod does not match the version used by the istio injector (pod version: %s; injector version: %s). This often happens after upgrading the Istio control-plane and can be fixed by redeploying the pod."
+    description: "The image of the Istio proxy running on the pod does not match the image defined in the injection configuration."
+    template: "The image of the Istio proxy running on the pod does not match the image defined in the injection configuration (pod image: %s; injection configuration image: %s). This often happens after upgrading the Istio control-plane and can be fixed by redeploying the pod."
     args:
-      - name: proxyVersion
+      - name: proxyImage
         type: string
-      - name: injectionVersion
+      - name: injectionImage
```
Comment (ayj) [inline galley/pkg/config/analysis/msg/messages.yaml]: @zerobfd, this changes the description and template text but not the schema itself. Any compat. concerns here?

Comment (jasonwzm) [reply galley/pkg/config/analysis/msg/messages.yaml]: Analysis message schema is not enforced and we are not currently generating schemas for the error and warning messages.

Comment (zerobfd) [reply galley/pkg/config/analysis/msg/messages.yaml]: LGTM.

**Pull request merged**